panic: "unexpected leftover memory" error during graceful shutdown

Seen on my local single-node instance, built at commit bfea8d95e15d0534b539584830ee3b62aa5ceccf. I used Ctrl+C to shutdown the instance, and while it was shutting down, it panic'ed:

```
I161101 17:53:12.782849 1035 cli/start.go:401  0 running tasks
W161101 17:53:12.870667 1034 cli/start.go:382  timed out waiting for 1 open connections to drain
I161101 17:53:12.870702 1034 util/stop/stopper.go:363  stop has been called, stopping or quiescing all running tasks
panic: unexpected leftover memory: 142 bytes [recovered]
	panic: unexpected leftover memory: 142 bytes

goroutine 112 [running]:
panic(0x50ec360, 0xc4205fc6d0)
	/usr/local/Cellar/go/1.7.1/libexec/src/runtime/panic.go:500 +0x1a1
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc4204a7050)
	/Users/alex/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:185 +0x6e
panic(0x50ec360, 0xc4205fc6d0)
	/usr/local/Cellar/go/1.7.1/libexec/src/runtime/panic.go:458 +0x243
github.com/cockroachdb/cockroach/pkg/sql/mon.(*MemoryUsageMonitor).StopMonitor(0xc42022d150, 0x9843168, 0xc42034fd80)
	/Users/alex/go/src/github.com/cockroachdb/cockroach/pkg/sql/mon/mem_usage.go:130 +0x2a2
github.com/cockroachdb/cockroach/pkg/sql.(*Session).Finish(0xc42022c900, 0xc4203d6a00)
	/Users/alex/go/src/github.com/cockroachdb/cockroach/pkg/sql/session.go:193 +0x99
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).finish(0xc42011c180, 0x9843000, 0xc42054f8c0)
	/Users/alex/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:154 +0x7e
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*Server).ServeConn(0xc42014aba0, 0x9843000, 0xc42054f8c0, 0x5ac2ea0, 0xc420412160, 0x5aab1a0, 0xc4205791d0)
	/Users/alex/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/server.go:247 +0x6bd
github.com/cockroachdb/cockroach/pkg/server.(*Server).Start.func8.1(0x5ac2ea0, 0xc420412160)
	/Users/alex/go/src/github.com/cockroachdb/cockroach/pkg/server/server.go:492 +0x106
github.com/cockroachdb/cockroach/pkg/util/netutil.(*Server).ServeWith.func1(0xc4204a7050, 0xc42012e268, 0x5ac2ea0, 0xc420412160, 0xc4202c8000)
	/Users/alex/go/src/github.com/cockroachdb/cockroach/pkg/util/netutil/net.go:136 +0x95
created by github.com/cockroachdb/cockroach/pkg/util/netutil.(*Server).ServeWith
	/Users/alex/go/src/github.com/cockroachdb/cockroach/pkg/util/netutil/net.go:138 +0x277
```

Ctrl+F doesn't find any log messages from the mem_usage.go file. @knz, is this just a side effect of not being able to drain the 1 open SQL connection or is it that already accounted for?

--- a/pkg/sql/create.go
+++ b/pkg/sql/create.go
@@ -616,11 +616,6 @@ func (n *createTableNode) Start() error {
 	}
 
 	if n.n.As() {
-		resultColumns := n.sourcePlan.Columns()
-		if err != nil {
-			return err
-		}
-
 		// TODO(knz): Ideally we would want to plug the sourcePlan which
 		// was already computed as a data source into the insertNode. Now
 		// unfortunately this is not so easy: when this point is reached,
@@ -631,12 +626,8 @@ func (n *createTableNode) Start() error {
 		n.sourcePlan.Close()
 		n.sourcePlan = nil
 
-		desiredTypesFromSelect := make([]parser.Type, len(resultColumns))
-		for i, col := range resultColumns {
-			desiredTypesFromSelect[i] = col.Typ
-		}
 		insert := &parser.Insert{Table: &n.n.Table, Rows: n.n.AsSource}
-		insertPlan, err := n.p.Insert(insert, desiredTypesFromSelect, false)
+		insertPlan, err := n.p.Insert(insert, nil, false)
 		if err != nil {
 			return err
 		}

--- a/pkg/sql/create.go
+++ b/pkg/sql/create.go
@@ -347,10 +347,10 @@ func (p *planner) CreateView(n *parser.CreateView) (planNode, error) {
 	if err != nil {
 		return nil, err
 	}
-
 	numColNames := len(n.ColumnNames)
 	numColumns := len(sourcePlan.Columns())
 	if numColNames != 0 && numColNames != numColumns {
+		sourcePlan.Close()
 		return nil, sqlbase.NewSyntaxError(fmt.Sprintf(
 			"CREATE VIEW specifies %d column name%s, but data source has %d column%s",
 			numColNames, util.Pluralize(int64(numColNames)),
@@ -358,10 +358,6 @@ func (p *planner) CreateView(n *parser.CreateView) (planNode, error) {
 	}
 
 	return &createViewNode{p: p, n: n, dbDesc: dbDesc, sourcePlan: sourcePlan}, nil
-}
-
-func (n *createViewNode) expandPlan() error {
-	return n.sourcePlan.expandPlan()
 }
 
 func (n *createViewNode) Start() error {
@@ -429,8 +425,13 @@ func (n *createViewNode) Start() error {
 	return nil
 }
 
+func (n *createViewNode) Close() {
+	n.sourcePlan.Close()
+	n.sourcePlan = nil
+}
+
+func (n *createViewNode) expandPlan() error                   { return n.sourcePlan.expandPlan() }
 func (n *createViewNode) Next() (bool, error)                 { return false, nil }
-func (n *createViewNode) Close()                              {}
 func (n *createViewNode) Columns() ResultColumns              { return make(ResultColumns, 0) }
 func (n *createViewNode) Ordering() orderingInfo              { return orderingInfo{} }
 func (n *createViewNode) Values() parser.DTuple               { return parser.DTuple{} }
@@ -446,7 +447,6 @@ type createTableNode struct {
 	p          *planner
 	n          *parser.CreateTable
 	dbDesc     *sqlbase.DatabaseDescriptor
-	insertPlan planNode
 	sourcePlan planNode
 }
 
@@ -491,6 +491,7 @@ func (p *planner) CreateTable(n *parser.CreateTable) (planNode, error) {
 		numColNames := len(n.AsColumnNames)
 		numColumns := len(sourcePlan.Columns())
 		if numColNames != 0 && numColNames != numColumns {
+			sourcePlan.Close()
 			return nil, sqlbase.NewSyntaxError(fmt.Sprintf(
 				"CREATE TABLE specifies %d column name%s, but data source has %d column%s",
 				numColNames, util.Pluralize(int64(numColNames)),
@@ -531,7 +532,7 @@ func hoistConstraints(n *parser.CreateTable) {
 }
 
 func (n *createTableNode) expandPlan() error {
-	if n.n.As() {
+	if n.sourcePlan != nil {
 		return n.sourcePlan.expandPlan()
 	}
 	return nil
@@ -619,10 +620,10 @@ func (n *createTableNode) Start() error {
 		// TODO(knz): Ideally we would want to plug the sourcePlan which
 		// was already computed as a data source into the insertNode. Now
 		// unfortunately this is not so easy: when this point is reached,
-		// sourcePlan.expandPlan() has already been called, and
-		// insertPlan.expandPlan() below would cause a 2nd invocation and
-		// cause a panic. So instead we close this sourcePlan and let the
-		// insertNode create it anew from the AsSource syntax node.
+		// sourcePlan.expandPlan() has already been called (for EXPLAIN),
+		// and insertPlan.expandPlan() below would cause a 2nd invocation
+		// and cause a panic. So instead we close this sourcePlan and let
+		// the insertNode create it anew from the AsSource syntax node.
 		n.sourcePlan.Close()
 		n.sourcePlan = nil
 
@@ -631,7 +632,7 @@ func (n *createTableNode) Start() error {
 		if err != nil {
 			return err
 		}
-		n.insertPlan = insertPlan
+		defer insertPlan.Close()
 		if err := insertPlan.expandPlan(); err != nil {
 			return err
 		}
@@ -651,8 +652,9 @@ func (n *createTableNode) Start() error {
 }
 
 func (n *createTableNode) Close() {
-	if n.insertPlan != nil {
-		n.insertPlan.Close()
+	if n.sourcePlan != nil {
+		n.sourcePlan.Close()
+		n.sourcePlan = nil
 	}
 }
 

--- a/pkg/sql/testdata/create_as
+++ b/pkg/sql/testdata/create_as
@@ -73,3 +73,10 @@ cups 10
 
 statement error pq: table "something" does not exist
 SELECT * FROM something LIMIT 1
+
+# Check for memory leak (#10466)
+statement ok
+CREATE TABLE foo (x, y, z) AS SELECT catalog_name, schema_name, sql_path FROM information_schema.schemata
+
+statement error pq: relation "foo" already exists
+CREATE TABLE foo (x, y, z) AS SELECT catalog_name, schema_name, sql_path FROM information_schema.schemata

--- a/pkg/sql/testdata/views
+++ b/pkg/sql/testdata/views
@@ -314,3 +314,10 @@ DROP VIEW s1;
 
 statement ok
 DROP TABLE t;
+
+# Check for memory leak (#10466)
+statement ok
+CREATE VIEW foo AS SELECT catalog_name, schema_name, sql_path FROM information_schema.schemata
+
+statement error pq: relation "foo" already exists
+CREATE VIEW foo AS SELECT catalog_name, schema_name, sql_path FROM information_schema.schemata

